src/lib.rs

```diff
@@ -488,6 +492,30 @@ pub trait Itertools : Iterator {
         groupbylazy::new_chunks(self, size)
     }
 
+    /// Return a `HashMap` whose keys are determined by a function applied to
+    /// each iterator item. The values of the HashMap are `Vec`s containing
+    /// the items which produced the corresponding key.
+    /// 
+    /// ```
+    /// use itertools::Itertools;
+    /// 
+    /// let data = vec![10, 12, 13, 20, 33, 42];
+    /// let lookup = data.into_iter().to_lookup(|&i| i % 10);
+    /// 
+    /// assert_eq!(lookup[&0], vec![10, 20]);
+    /// assert_eq!(lookup.get(&1), None);
+    /// assert_eq!(lookup[&2], vec![12, 42]);
+    /// assert_eq!(lookup[&3], vec![13, 33]);
+    /// ```
+    #[cfg(feature = "use_std")]
+    fn to_lookup<K, F>(self, get_key: F) -> HashMap<K, Vec<Self::Item>>
+        where Self: Iterator + Sized,
+              K: Hash + Eq,
+              F: Fn(&Self::Item) -> K
+    {
```

Comment:
Naming aside (it is important too).

I would expect this to apply to iterators of pairs and use the first field as the key and the second as the value. Like `HashMap::from_iter` expects it. Then we don't have to worry about the diversity in calling mode required to extracting the key, be it by value (split the key out as a part of something) or by ref (clone/create a new key from a value).

Reply:
I was actually thinking about just this :) haven't had much time to work on it this weekend but I was considering providing three different argument signatures:
* iterator of pairs,
* iterator with key function (i.e. the current version),
* iterator with both key and value functions.

Seems reasonable enough to only provide the first one. However, my initial reasoning for doing it as it currently is, was to match `.group_by()`. I can't see any reason why this function and that one would take different arguments. Maybe `.group_by()` should take a pair iterator too?

Reply:
Good point, they seem to be in about the same situation. Maybe they differ in that the group by key is often not kept around.

It seems they should both do the same thing, and if I had thought of that, I'd have picked to mimic group_by as well.